accepted/2021/preview-features/preview-features.md

```diff
@@ -0,0 +1,418 @@
+behavioral changes and also allows certain parts of .NET to make deliberate
+breaking changes if necessary.
+
+As a result, this makes it viable to innovate across runtime, libraries, and
+languages, allowing us to build more compelling features that would have been
+almost impossible to do on .NET Framework. Examples include static linking,
+default interface members, and ref structs (such as `Span<T>`).
+
+However, it's sometimes challenging to deliver cross-cutting feature work in a
+single release, especially when they are breaking new ground in terms of
+expressiveness. A good example of such a feature is the upcoming improved
+numeric support in generic code, with the help of static interface members. Such
+a feature requires work in the runtimes, the languages, and the libraries. The
+work at the library level isn't very high, but it's all about finding the right
+set of support infrastructure so that developers are able to be productive at
+writing generic code that can rely on numerical expressions and operators. In a
+sense, the library work can't really fully start until the runtime and compiler
+support is sufficiently bootstrapped, but it's the most important part in the
+validation of the feature set. Practically, this means a good chunk of the
+feature will only be available halfway through a release cycle, which greatly
+reduces the time window for gathering feedback from other parts of the product
+as well as from customers using preview builds of the upcoming .NET version.
+
+Sadly, preview feedback isn't always sufficient to judge how well a feature
+works because
+
+1. Many fewer people use preview builds of .NET and
+2. Many people don't use .NET previews for their production projects, thus
+   greatly reducing the breadth & depth at which previews are being exercised.
+
+This document is exploring a model where GA quality builds of .NET can contain
+features that are considered still in preview.
+
+The goal is to ensure that customers can discover those features & have an easy
+way to play with them while simultaneously being fully aware that these features
+aren't supported yet and thus might change (or be removed entirely) between
+versions of .NET, just like they could in preview builds.
+
+## Scenarios and User Experience
+
+### Enabling a preview runtime feature
+
+Maddy is building a Tye application. She reads in a blog post that the .NET
+runtime team is experimenting with a new GC mode, code named FancyGC.
+
+She wants to see how this improves the performance of her backend. She's glad to
+see that the new GC mode was shipped as part of .NET 6, so she doesn't need to
+figure out how to run a new version of .NET in her CI lab. So she creates a
+feature branch and enables the new GC mode simply by changing the service's
+project file as follows:
+
+```xml
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>net6.0</TargetFramework>
+    <EnablePreviewFeatures>True</EnablePreviewFeatures>
```

Comment:
Why is this separate from langversion?  Why can't we just use the existing `<LangVersion>preview</LangVersion>` to mean "I'm opting in to preview stuff"?  I don't think many folks realistically differentiate between language, library, and runtime, nor should they need to?  From my perspective, we already have a mechanism for this... we should just use it.

Reply:
It would also be easier to communicate.

Reply:
Because LangVersion represents the language version. To me, it doesn't make sense to say that setting `LangVersion` to preview means that it also enables runtime and framework features.

I'm suggesting that `EnablePreviewFeatures` is the single setting for most users as we generally don't want users to set `LangVersion` anyway.

Reply:
> To me, it doesn't make sense to say that setting LangVersion to preview means that it also enables runtime and framework features.

It enables the full stack preview features.  I think we should continue using the thing we already have.

Sounds like we need another discussion on this point.

Reply:
It sure does 🙂

FWIW, I'm not opposed to using `LangVersion` for this if we believe that's the right setting. My understanding from talking to @MadsTorgersen and @jaredpar was that we generally don't want people to set `LangVersion` in the project file.

Reply:
My biggest concern is that not all of our cross cutting work will include language support. Consider hardware intrinsics as a concrete example. That is a cross cutting feature but had no language involvement. Hence seems odd to have `<LangVersion>preview</LangVersion>` as the way to opt into that. I don't have a strong objection to using `LangVersion` but that is one oddity that comes to mind. 

I do feel strongly that we should have one way of opting into cross cutting previews. By that I mean if we ended up picking say `<EnablePreviewFeatures>true</EnablePreviewFeatures>` then I would feel strongly that `<LangVersion>` should be defalut to `"preview"` when that was enabled.

Reply:
For now, we landed on having the `EnablePreviewFeatures` as the primary switch which also sets `LangVersion` to `Preview` (unless the user explicitly set it).